### PR TITLE
Development- Get Endpoints for instance and binding

### PR DESCRIPTION
**Pull request by anujbansal16**
**HCPCFS-2936**
#### As per the changes introduced by [Openservicebrokerapi](https://github.com/openservicebrokerapi/servicebroker/blob/v2.14/spec.md#changes-since-v213) (since  v2.13) we enhanced our broker to support following two GET endpoints. Changes are listed below:
- Added GET endpoints for fetching a [Service Instance.](https://github.com/openservicebrokerapi/servicebroker/blob/v2.14/spec.md#fetching-a-service-instance)

- Added GET endpoints for fetching a [Service Binding.](https://github.com/openservicebrokerapi/servicebroker/blob/v2.14/spec.md#fetching-a-service-binding)

These endpoints requires at least 2.14 as API version. i.e. `X-Broker-API-Version: 2.14`

---------------------------------------------------------------------------------------------------
**HCPCFS-2939**

**Requirement**: A Platform might wish to uniquely identify a specific request as it flows throughout the system. For example, this might be used for logging for request tracing purposes. In order to facilitate this, Platforms will need to provide identification information to the Service Broker for each request. Platforms MAY support this feature, and if they do, they MUST adhere to the following:
- For any OSBAPI request, there MUST be an associated X-Broker-API-Request-Identity header on the HTTP request.
- The Service Broker MAY include this value in log messages generated as a result of the request.
- The Service Broker SHOULD include this header in the response to the request. The format of the header MUST be:
`X-Broker-API-Request-Identity: value`
value MUST be a non-empty string indicating the identity of the request being sent. The specific value MAY be unique for each request sent to the broker. Using a GUID is RECOMMENDED.

**Changes**
- Added [Request Identity Header.](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#request-identity) The Service Broker SHOULD include this header in the response to the request. i.e.
`X-Broker-API-Request-Identity: value`. [(Commit)](https://github.com/cloudfoundry-incubator/service-fabrik-broker/pull/1239/commits/e79af3283cbc9e3cf667ec7d679a82d8ee8456ff)

### CODE REVIEW

File: broker/applications/osb-broker/src/api-controllers/ServiceBrokerApiController.js [hunk 1]
```diff
@@ -519,6 +519,136 @@ class ServiceBrokerApiController extends FabrikBaseController {
...
+          if(!isServiceBindingRetrievable) {
+            throw new BadRequest('Service does not support binding retrieval');
+          }
+          const isPlanBindable = _.get(catalog.getPlan(_.get(resource, 'spec.planId')), 'bindable',isServiceBindable);
+          // if plan is bindable
+          if(!isPlanBindable) {
+            throw new BadRequest('Service plan is not bindable');
+          }
+          const resourceStatus = _.get(resource, 'status');
+          const resourceState = resourceStatus.state;
```
Comment (jayeshjk) [inline broker/applications/osb-broker/src/api-controllers/ServiceBrokerApiController.js:648]: This can be done in single `_.get` call

**Review by jayeshjk — CHANGES_REQUESTED** [review]
Few comments towards making code a little bit concise.

File: broker/applications/osb-broker/src/api-controllers/ServiceBrokerApiController.js [hunk 2]
```diff
@@ -140,7 +140,8 @@ class ServiceBrokerApiController extends FabrikBaseController {
         spec: params,
         status: {
           state: CONST.APISERVER.RESOURCE_STATE.IN_QUEUE
-        }
+        },
+        requestIdentity: _.get(req.headers, 'x-broker-api-request-identity')
```
Comment (jayeshjk) [inline broker/applications/osb-broker/src/api-controllers/ServiceBrokerApiController.js:144]: Read `x-broker-api-request-identity` from constants. (Please do this for other occurrences as well)

File: broker/applications/osb-broker/src/api-controllers/ServiceBrokerApiController.js [hunk 3]
```diff
@@ -370,7 +378,8 @@ class ServiceBrokerApiController extends FabrikBaseController {
       resourceGroup: resourceGroup,
       resourceType: resourceType,
       resourceId: resourceId,
-      namespaceId: resourceType === CONST.APISERVER.RESOURCE_TYPES.INTEROPERATOR_SERVICEINSTANCES ? eventmesh.apiServerClient.getNamespaceId(resourceId) : undefined
+      namespaceId: resourceType === CONST.APISERVER.RESOURCE_TYPES.INTEROPERATOR_SERVICEINSTANCES ? eventmesh.apiServerClient.getNamespaceId(resourceId) : undefined,
```
Comment (jayeshjk) [inline broker/applications/osb-broker/src/api-controllers/ServiceBrokerApiController.js:381]: Please check if `getKubernetesName` is needed here.

File: broker/applications/osb-broker/src/api-controllers/ServiceBrokerApiController.js (hunk 2 above)
Comment (anujbansal16) [reply broker/applications/osb-broker/src/api-controllers/ServiceBrokerApiController.js]: Done[(Commit)](https://github.com/cloudfoundry-incubator/service-fabrik-broker/pull/1239/commits/86bfb2688e355784cb6595874498add0dc82c187)

File: broker/applications/osb-broker/src/api-controllers/ServiceBrokerApiController.js [hunk 4]
```diff
@@ -519,6 +519,136 @@ class ServiceBrokerApiController extends FabrikBaseController {
...
+          } else if(resourceState === CONST.APISERVER.RESOURCE_STATE.UPDATE) {
+            // resource is being updated?
+            throw new UnprocessableEntity('Service Instance is being updated and therefore cannot be fetched at this time', 'ConcurrencyError');
+          } else if(resourceState === CONST.APISERVER.RESOURCE_STATE.DELETE) {
+            // resource is being deleted?
+            throw new UnprocessableEntity('Service Instance is being deleted and therefore cannot be fetched at this time', 'ConcurrencyError');
+          } else if(resourceState === CONST.OPERATION.IN_PROGRESS || resourceState === CONST.APISERVER.RESOURCE_STATE.IN_PROGRESS) {
+            // check the last operation and send 422 accordingly.
+            const labels = _.get(resource, 'metadata.labels', false);
+            const lastOperation = labels ? labels['interoperator.servicefabrik.io/lastoperation'] : '';
```
[the remark this answers is not in the document]
Comment (anujbansal16) [reply broker/applications/osb-broker/src/api-controllers/ServiceBrokerApiController.js]: @jayeshjk Done.[(Commit)](https://github.com/cloudfoundry-incubator/service-fabrik-broker/pull/1239/commits/f053553724dea3d6efb9638724736208a7d821da)

File: broker/applications/osb-broker/src/api-controllers/ServiceBrokerApiController.js (hunk 1 above)
Comment (anujbansal16) [reply broker/applications/osb-broker/src/api-controllers/ServiceBrokerApiController.js]: @jayeshjk Done. [(Commit)](https://github.com/cloudfoundry-incubator/service-fabrik-broker/pull/1239/commits/6c863f9d5ddda71cf052b9a38c70b31e870421ba)

File: broker/applications/osb-broker/test/acceptance/service-broker-api.GetEndpoints.spec.js [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+const config = require('@sf/app-config');
+const {
+  CONST,
+  commonFunctions: {
+    encodeBase64
+  }
+} = require('@sf/common-utils');
+const camelcaseKeys = require('camelcase-keys');
+
+describe('service-broker-api-enhancement', function () {
```
Comment (jayeshjk) [inline broker/applications/osb-broker/test/acceptance/service-broker-api.GetEndpoints.spec.js:16]: Can you pls follow the hierarchy in [service-broker-api-2.0.instances.director.spec.js](https://github.com/cloudfoundry-incubator/service-fabrik-broker/blob/9aaa417e70f5237bdcb80f65b8c0dd42e67a75e3/broker/applications/osb-broker/test/acceptance/service-broker-api-2.0.instances.director.spec.js) . For example, hierarchical position for tests for provision operation is: `describe('service-broker-api-2.0') --> describe('director') --> describe('#provision')`

File: broker/applications/osb-broker/src/api-controllers/middleware/index.js [hunk 6]
```diff
@@ -202,4 +203,24 @@ exports.checkBlockingOperationInProgress = function () {
...
+  return function (req, res, next) {
+    const version = _.get(req.headers, 'x-broker-api-version', '1.0');
+    if(commonFunctions.compareVersions(version, minVersion) < 0) {
+      return next(new PreconditionFailed(`At least Broker API version ${minVersion} is required.`));
+    }
+    next();
+  };
+};
+
+exports.addRequestIdentity = function () {
```
Comment (jayeshjk) [inline broker/applications/osb-broker/src/api-controllers/middleware/index.js:216]: Does `addRequestIdentityToResponse` sound better?

**Review by jayeshjk — CHANGES_REQUESTED** [review]
Few more minor comments.

File: broker/applications/osb-broker/src/api-controllers/middleware/index.js (hunk 6 above)
Comment (anujbansal16) [reply broker/applications/osb-broker/src/api-controllers/middleware/index.js]: @jayeshjk  Done

File: broker/applications/osb-broker/test/acceptance/service-broker-api.GetEndpoints.spec.js (hunk 5 above)
Comment (anujbansal16) [reply broker/applications/osb-broker/test/acceptance/service-broker-api.GetEndpoints.spec.js]: @jayeshjk Done.

**Comment by vinaybheri** [comment]
Validation on k8s cluster: n-2 is succeeded.

**Comment by vinaybheri** [comment]
Validation on k8s cluster: n-1 is succeeded.

**Comment by vinaybheri** [comment]
Validation on k8s cluster: n is succeeded.

**Review by jayeshjk — APPROVED** [review]
LGTM

**Comment by vinaybheri** [comment]
Validation on k8s cluster: n-1 is succeeded.

**Comment by vinaybheri** [comment]
Validation on k8s cluster: n is succeeded.

**Comment by vinaybheri** [comment]
Validation on k8s cluster: n is succeeded.

**Review by vivekzhere — APPROVED** [review]
Looks good

**Pull request merged**